src/raft.rs

```diff
@@ -1039,7 +1152,123 @@ impl<T: Storage> Raft<T> {
                 StateRole::Leader => self.step_leader(m)?,
             },
         }
+        Ok(())
+    }
 
+    /// Apply a `BeginConfChange` entry.
+    ///
+    /// When a Raft node applies this variant of a configuration change it will adopt a joint
+    /// configuration state until the membership change is finalized.
+    ///
+    /// During this time the `Raft` will have two, possibly overlapping, cooperating quorums for
+    /// both elections and log replication.
+    ///
+    /// # Implementation notes
+    ///
+    /// This uses a slightly modified "Joint Consensus" algorithm as detailed in Section 6 of the
+    /// Raft paper.
+    ///
+    /// We apply the change when a node *applies* the entry, not when the entry is received.
+    ///
+    /// # Contracts
+    ///
+    /// * The `entry: Entry` is of type `EntryType::ConfChange` and the `data` field is a
+    /// `ConfChange` serialized by `protobuf::Message::write_to_bytes`.
+    /// * The `ConfChange` must be of type `BeginConfChange` and contain a `configuration` value.
+    #[inline(always)]
+    pub fn begin_membership_change(&mut self, entry: &Entry) -> Result<()> {
```

Comment:
Personally I prefer `ConfChange` instead of `Entry`. One reason is about style: `add_learner` and `and_node` didn't use `Entry`. And the another reason is if the method is called out of this crate, the `Entry` has been already parsed generally.

Reply:
Part of the reason for this is so we can do debug assertions to make sure the log index (found in the entry) actually contains the given ConfChange.